app/javascript/angular/dashboard/widget-wrapper.js

```diff
@@ -26,23 +25,49 @@ ManageIQ.angular.app.component('widgetWrapper', {
     vm.$onInit = function() {
       vm.divId = `w_${vm.widgetId}`;
       vm.innerDivId = `dd_w${vm.widgetId}_box`;
+      vm.refreshWidgetHTML(false);
+      vm.parsedButtons = JSON.parse(vm.widgetButtons);
+      const refreshButton = vm.parsedButtons.find(ob => ob.id.includes("refresh"));
+      if (refreshButton) {
+        refreshButton.onclick = vm.refresh;
+      };
+    };
 
-      if (vm.widgetBlank === 'false') {
-        $http.get(vm.widgetUrl())
-          .then((response) => {
-            vm.widgetModel = response.data;
-            // if there's html make it passable
-            if (vm.widgetModel.content) {
-              vm.widgetModel.content = $sce.trustAsHtml(vm.widgetModel.content);
-            }
-            deferred.resolve();
-          })
-          .catch((e) => {
-            vm.error = true;
-            miqService.handleFailure(e);
-            deferred.reject();
-          });
-      }
+    vm.refreshWidgetHTML = function(refreshed) {
+      return $http.get(vm.widgetUrl())
+        .then((response) => {
+          vm.widgetModel = response.data;
+          // if there's html make it passable
+          if (vm.widgetModel.content) {
+            vm.widgetModel.content = $sce.trustAsHtml(vm.widgetModel.content);
+          }
+          if (refreshed) {
+            miqSparkleOff();
+            add_flash(sprintf(__('Dashboard "%s" was refreshed'), vm.widgetTitle), 'success');
+          }
+          vm.error = false;
+          deferred.resolve();
+        })
+        .catch((e) => {
+          vm.error = true;
+          vm.widgetModel = null;
+          miqService.handleFailure(e);
+          deferred.reject();
+        });
+    };
+
+    vm.refresh = function() {
+      $http.post(`/dashboard/widget_refresh/?widget=${vm.widgetId}`)
+        .then((response) => {
+          vm.widgetModel = null;
+          return API.wait_for_task(response.data.task_id).then(vm.refreshWidgetHTML(true));
```

Comment:
you're calling the function right away now, not actually waiting for the promise

`() => vm.refreshWidgetHTML(true)`?

Reply:
Fixed.